### ISSUE TITLE
NoClassDefFoundError when exception occurs inside a Coroutine in jlink generated runtime

### ISSUE THREAD

**Issue by JonathanxD** [Kotlin/kotlinx.coroutines#3032]
I've been trying to create a runtime for my JavaFX application based upon the Gradle file generated by IntelliJ from the following settings:

![image](https://user-images.githubusercontent.com/5360060/142475337-ffac91ff-61ec-4401-aa98-8058ef44cec6.png)
![image](https://user-images.githubusercontent.com/5360060/142475247-b074ab18-9dba-4a07-b3cd-d390c965206f.png)

I needed to tweak the generated `build.gradle` a bit to get Kotlin, Kotlinx Serialization and Kotlinx Coroutines to work, here is my `build.gradle`:
```
plugins {
    id 'java'
    id 'application'
    id 'org.jetbrains.kotlin.jvm' version '1.6.0'
    id 'org.jetbrains.kotlin.plugin.serialization' version '1.6.0'
    id 'org.openjfx.javafxplugin' version '0.0.10'
    id 'org.beryx.jlink' version '2.24.4'
}

group 'com.github.jonathanxd'
version '1.0-SNAPSHOT'

repositories {
    mavenCentral()
}

ext {
    junitVersion = '5.8.1'
}

sourceCompatibility = '16'
targetCompatibility = '16'

tasks.withType(JavaCompile) {
    options.encoding = 'UTF-8'
}

application {
    //mainModule = 'com.github.jonathanxd.modinstaller'
    mainClass = 'com.github.jonathanxd.modinstaller.HelloApplication'
}

[compileKotlin, compileTestKotlin].forEach {
    it.kotlinOptions {
        jvmTarget = '16'
    }
}

javafx {
    version = '16'
    modules = ['javafx.base', 'javafx.graphics', 'javafx.controls', 'javafx.fxml', 'javafx.web']
}

dependencies {
    implementation 'net.sf.sevenzipjbinding:sevenzipjbinding:16.02-2.01'
    implementation 'net.sf.sevenzipjbinding:sevenzipjbinding-all-platforms:16.02-2.01'

    implementation('org.controlsfx:controlsfx:11.1.0')
    implementation('com.dlsc.formsfx:formsfx-core:11.3.2') {
        exclude(group: 'org.openjfx')
    }
    implementation('net.synedra:validatorfx:0.1.13') {
        exclude(group: 'org.openjfx')
    }
    implementation('org.kordamp.ikonli:ikonli-javafx:12.2.0')
    implementation('org.kordamp.bootstrapfx:bootstrapfx-core:0.4.0')
    implementation('eu.hansolo:tilesfx:11.48') {
        exclude(group: 'org.openjfx')
    }
    implementation 'org.openjfx:javafx:16'
    implementation 'org.openjfx:javafx-controls:16'
    implementation 'org.openjfx:javafx-graphics:16'
    implementation("org.jetbrains.kotlinx:kotlinx-serialization-json:1.3.1")
    implementation 'org.jetbrains.kotlinx:kotlinx-coroutines-core:1.5.2'
    implementation 'org.jetbrains.kotlin:kotlin-stdlib:1.6.0'
    implementation 'org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.6.0'

    testImplementation("org.junit.jupiter:junit-jupiter-api:${junitVersion}")
    testRuntimeOnly("org.junit.jupiter:junit-jupiter-engine:${junitVersion}")
}

test {
    useJUnitPlatform()
}

jlink {
    imageZip = project.file("${buildDir}/distributions/app-${javafx.platform.classifier}.zip")
    options = ['--strip-debug', '--compress', '2', '--no-header-files', '--no-man-pages']
    forceMerge("kotlin-stdlib", "kotlin-stdlib-common", "kotlin-stdlib-jdk7", "kotlin-stdlib-jdk8", "kotlinx-coroutines-core-jvm", "kotlinx-serialization-core-jvm", "kotlinx-serialization-json-jvm")
    launcher {
        name = 'app'
    }
}

jlinkZip {
    group = 'distribution'
}
```

Everything works fine, however, when an exception occurs inside a Coroutine, more specifically when the exception happens inside a `withContext(Dispatchers.IO) { ... }`, the following error occurs, instead of regular exception handling:
```
Exception in thread "DefaultDispatcher-worker-23" java.lang.NoClassDefFoundError: Could not initialize class kotlinx.coroutines.CoroutineExceptionHandlerImplKt
        at com.github.jonathanxd.merged.module@1.0-SNAPSHOT/kotlinx.coroutines.CoroutineExceptionHandlerKt.handleCoroutineException(Unknown Source)
        at com.github.jonathanxd.merged.module@1.0-SNAPSHOT/kotlinx.coroutines.DispatchedTask.handleFatalException(Unknown Source)
        at com.github.jonathanxd.merged.module@1.0-SNAPSHOT/kotlinx.coroutines.DispatchedTask.run(Unknown Source)
        at com.github.jonathanxd.merged.module@1.0-SNAPSHOT/kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(Unknown Source)
        at com.github.jonathanxd.merged.module@1.0-SNAPSHOT/kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(Unknown Source)
        at com.github.jonathanxd.merged.module@1.0-SNAPSHOT/kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(Unknown Source)
        at com.github.jonathanxd.merged.module@1.0-SNAPSHOT/kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(Unknown Source)
```

The current workaround is to manually add a `try-catch` and handle the exception on my own, instead of delegating this job to the Coroutine engine, which seems to be counter-intuitive.

I'm not sure if this problem is on the Kotlinx Coroutines end, but assuming that this exception was seen before and reported in #1300, I'm bringing this problem here

**Comment by chxchen** [comment]
i got a similar error. change kotlin-corountines version to 1.4.3, it works fine.

**Comment by dosier** [comment]
Same issue here using
```kt
    val ktor_version = "2.1.1"
    val coroutine_version = "1.6.4"
 ```

**Comment by qwwdfsad** [comment]
Should be fixed with 1.7.0-Beta where proper JPMS descriptor are packaged into JARs

**Comment by laurentiu-git** [comment]
> Should be fixed with 1.7.0-Beta where proper JPMS descriptor are packaged into JARs

I am still getting this issue using the 1.7.0-Beta version and I am doing something similar to what @dosier is doing and is mentioned in [218](https://github.com/beryx/badass-jlink-plugin/issues/218)
